### PR TITLE
Storage and Environment Big Refactor

### PR DESCRIPTION
**Pull request by joshmeek**
**Thanks for contributing to Prefect!**

Please describe your work and make sure your PR:

- [x] adds new tests (if appropriate)
- [x] updates `CHANGELOG.md` (if appropriate)
- [x] updates docstrings for any new functions or function arguments, including `docs/outline.toml` for API reference docs (if appropriate)

Note that your PR will not be reviewed unless all three boxes are checked.

This is being merged into `dev` due to pseudo-breaking changes.

## What does this PR change?
Adds new storage and environment separation to Prefect.

## Why is this PR important?
Cleaner way to store, share, and run flows in a distributed fashion.

### CODE REVIEW

**Comment by jlowin** [comment]
:clap: test coverage > 95%

File: src/prefect/cli/__init__.py [hunk 1]
```diff
@@ -19,27 +19,15 @@ def cli():
...
+@click.argument("storage_metadata")
 @click.argument("environment_metadata")
-def create_environment(environment_metadata):
-    """
-    Call the setup and execute functions for a given environment.
-    """
-    schema = prefect.serialization.environment.EnvironmentSchema()
-    environment = schema.load(json.loads(environment_metadata))
+def execute_flow(storage_metadata, environment_metadata):
+    """"""
```
Comment (cicdw) [inline src/prefect/cli/__init__.py]: Doesn't need to be for the merge into dev, but could you add a doc for this?  I'm curious about how complicated of a call this requires.

File: src/prefect/core/flow.py [hunk 2]
```diff
@@ -106,7 +107,9 @@ def my_task():
         - name (str): The name of the flow. Cannot be `None` or an empty string
         - schedule (prefect.schedules.Schedule, optional): A default schedule for the flow
         - environment (prefect.environments.Environment, optional): The environment
-            type that the flow should be run in. If None, a LocalEnvironment will be created.
+            type that the flow should be run in. If None, a CloudEnvironment will be created.
```
Comment (cicdw) [inline src/prefect/core/flow.py]:
```suggestion
           that the flow should be run in. If `None`, a `CloudEnvironment` will be created.
```

File: src/prefect/core/flow.py [hunk 3]
```diff
@@ -156,7 +160,8 @@ def __init__(
 
         self.name = name
         self.schedule = schedule
-        self.environment = environment or prefect.environments.LocalEnvironment()
+        self.environment = environment or prefect.environments.CloudEnvironment()
```
Comment (cicdw) [inline src/prefect/core/flow.py]: 🔥

File: src/prefect/environments/execution/cloud/environment.py [hunk 4]
```diff
@@ -0,0 +1,200 @@
...
+                registry_url=registry_url,
+                image_name=image_name,
+                image_tag=image_tag,
+                flow_file_path=flow_file_path,
+            )
+
+            # Create Job
+            batch_client.create_namespaced_job(namespace="default", body=job)
+
+    def run_flow(self) -> None:
```
Comment (cicdw) [inline src/prefect/environments/execution/cloud/environment.py]: Should this be changed to `run()` to match command that you submit in the `yml` file below?

File: src/prefect/environments/storage/bytes.py [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+        Build the Bytes storage object.
+
+        Args:
+            - flow (prefect.Flow): Flow to be stored
+
+        Returns:
+            - Bytes: a new Bytes storage object that contains information about how and
+                where the flow is stored.
+        """
+        return Bytes()
```
Comment (cicdw) [inline src/prefect/environments/storage/bytes.py]:
```
# TODO: implement this
```
?

File: src/prefect/cli/__init__.py (hunk 1 above)
Comment (joshmeek) [reply src/prefect/cli/__init__.py]: I'll add it to the list of issues I want to make once this is merged in

File: src/prefect/environments/storage/bytes.py (hunk 5 above)
Comment (joshmeek) [reply src/prefect/environments/storage/bytes.py]: This is also one of the issues I will be opening

**Comment by joshmeek** [comment]
@cicdw Not sure why 3.7 failed 

```
= 3295 passed, 13 skipped, 2 xfailed, 28 xpassed, 6 warnings in 76.25 seconds ==
Too long with no output (exceeded 10m0s)
````

**Pull request merged**